### PR TITLE
Lecture4 - new PR

### PR DESCRIPTION
**Pull request by VictorLVA**

### CODE REVIEW

File: Lecture4/Task4/Task4L4.java [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+                        Voucher newVoucher = VoucherFactory.getVoucher(voucherDetails);
+                        listOfVouchers.add(newVoucher);
+                    } else {
+                        notDownloadedCount++;
+                    }
+                } catch (NumberFormatException
+                        | ArithmeticException
+                        | ArrayIndexOutOfBoundsException
+                        | RegexException initEx) {
+                    notDownloadedCount++;
```
Comment (dihnatsyeu) [inline Lecture4/Task4/Task4L4.java:65]: when you catch any exception, good practice to log this event. For now, it's enough just to print error message to console.

File: Lecture4/Task4/VoucherFactory.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import predefinedExceptions.RegexException;
+
+class VoucherFactory {
+    static Voucher getVoucher(String[] voucherDetails) throws RegexException {
+        switch (Integer.parseInt(voucherDetails[0])) {
+            case 1:
+                return new RestVoucher(voucherDetails);
```
Comment (dihnatsyeu) [inline Lecture4/Task4/VoucherFactory.java:7]: parse voucher details here, not in constructors. Constructors should be protected from structural changes in file

File: Lecture4/Task4/Voucher.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    private String startDate;
+    private int countDays;
+    private int discount;
+    private double cost;
+    private String formattedCost;
+
+    Voucher(String[] voucherDetails) throws RegexException {
+        this.name = voucherDetails[1];
+        this.price = Double.parseDouble(voucherDetails[2]);
+        if (voucherDetails[3].matches("[A-Z][A-Z]|[A-Z][A-Z][A-Z]")) {
```
Comment (dihnatsyeu) [inline Lecture4/Task4/Voucher.java:19]: 1)don't put too much code to constructor.
2) every if-else block can be extracted to a separate method
3) Complex boolean conditions in if, such like this "voucherDetails[4].matches("(0?[1-9]|[12][0-9]|3[01])[.](0?[1-9]|1[012])[.]\\d{4}")" can be extracted to private method like isDate(). The code will be reformatted to if (isDate()) {...} - much simplier
4) Split responsibilities - read file and parse it context in different classes(methods).

File: Lecture4/Task4/VoucherFactory.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+            case 2:
+                return new CureVoucher(voucherDetails);
+            case 3:
+                return new ShoppingVoucher(voucherDetails);
+        }
+        return null;
```
Comment (dihnatsyeu) [inline Lecture4/Task4/VoucherFactory.java:13]: don't return null. Instead define default behavior to throw custom exception if type of voucher is not recognized.

File: Lecture4/Task4/predefinedExceptions/SelectedVoucherTypeException.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+    private String message;
+
+    public SelectedVoucherTypeException(String message) {
+        this.message = message;
+    }
+
+    @Override
+    public String getMessage() {
+        return message;
```
Comment (dihnatsyeu) [inline Lecture4/Task4/predefinedExceptions/SelectedVoucherTypeException.java:13]: why do you think you need to override default getMessage()?

File: Lecture4/Tasks1-3/Task1_3L4.java [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+                        | ArrayIndexOutOfBoundsException
+                        | RegexException initEx) {
+                    notDownloadedCount++;
+                }
+            }
+        } catch (IOException inputEx) {
+            inputEx.printStackTrace();
+            System.exit(0);
+        }
+        if (notDownloadedCount != 0) {
```
Comment (dihnatsyeu) [inline Lecture4/Tasks1-3/Task1_3L4.java:82]: 82-83 lines you can put to finally block if you will face not catch exceptions

File: Lecture4/Tasks1-3/Task1_3L4.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+                userInput = input.nextInt();
+            } catch (InputMismatchException inputMismatchEx) {
+                System.out.print(MESSAGE_TRY_AGAIN);
+                continue;
+            }
+            try {
+                if (userInput >= 1 && userInput <= 3) {
+                    selection = userInput;
+                } else {
+                    throw new SelectedVoucherTypeException(MESSAGE_TRY_AGAIN);
```
Comment (dihnatsyeu) [inline Lecture4/Tasks1-3/Task1_3L4.java:103]: throwing and catching the same exception make no sense in the same code block

File: Lecture4/Task4/Voucher.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            this.country = voucherDetails[3];
+        } else {
+            throw new RegexException();
```
Comment (dihnatsyeu) [inline Lecture4/Task4/Voucher.java:22]: bad tone to throw an exception without details. You didn't provide any error message

File: Lecture4/Task4/Task4L4.java [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            }
+        } catch (EmptyListException emptyListEx) {
+            emptyListEx.getMessageAndExit();
+        }
+    }
+
+    private static void writeSelectedToFile(Voucher selectedVoucher) {
+        String[] voucherDetails = selectedVoucher.toString().split(" \\| ");
+        try {
+            FileWriter voucherDetailsWriter = new FileWriter("Your voucher.txt");
```
Comment (dihnatsyeu) [inline Lecture4/Task4/Task4L4.java:296]: in additional to writing voucher as string, implement a method to write voucher as object

Comment (VictorLVA) [reply Lecture4/Task4/Task4L4.java:296]: fixed in Tasks1-4-fixed package

File: Lecture4/Tasks1-3/Task1_3L4.java (hunk 7 above)
Comment (VictorLVA) [reply Lecture4/Tasks1-3/Task1_3L4.java:103]: fixed in Tasks1-4-fixed package

File: Lecture4/Tasks1-3/Task1_3L4.java (hunk 6 above)
Comment (VictorLVA) [reply Lecture4/Tasks1-3/Task1_3L4.java:82]: fixed in Tasks1-4-fixed package

File: Lecture4/Task4/predefinedExceptions/SelectedVoucherTypeException.java (hunk 5 above)
Comment (VictorLVA) [reply Lecture4/Task4/predefinedExceptions/SelectedVoucherTypeException.java:13]: fixed indirectly in Tasks1-4-fixed package

File: Lecture4/Task4/Voucher.java (hunk 3 above)
Comment (VictorLVA) [reply Lecture4/Task4/Voucher.java:19]: fixed in Tasks1-4-fixed package

File: Lecture4/Task4/VoucherFactory.java (hunk 2 above)
Comment (VictorLVA) [reply Lecture4/Task4/VoucherFactory.java:7]: fixed in Tasks1-4-fixed package

File: Lecture4/Task4/Task4L4.java (hunk 1 above)
Comment (VictorLVA) [reply Lecture4/Task4/Task4L4.java:65]: fixed in Tasks1-4-fixed package

File: Lecture4/Task4/Voucher.java (hunk 8 above)
Comment (VictorLVA) [reply Lecture4/Task4/Voucher.java:22]: fixed in Tasks1-4-fixed package
Error message will be provided when exception will be catched and processed

File: Lecture4/Task4/VoucherFactory.java (hunk 4 above)
Comment (VictorLVA) [reply Lecture4/Task4/VoucherFactory.java:13]: fixed in Tasks1-4-fixed package

**Pull request merged**